vkcs/kubernetes/shared.go

```diff
@@ -103,3 +103,59 @@ func kubernetesStateRefreshFunc(client *gophercloud.ServiceClient, clusterID str
 		return c, c.NewStatus, nil
 	}
 }
+
+func kubernetesNodeGroupCreateStateRefreshFunc(client *gophercloud.ServiceClient, nodeGroupID string, clusterID string) retry.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		n, err := nodegroups.Get(client, nodeGroupID).Extract()
+		if err != nil {
+			if errutil.IsNotFound(err) {
+				return n, string(nodeGroupStatusCreating), nil
+			}
+			return nil, "", err
+		}
+		c, err := clusters.Get(client, clusterID).Extract()
+		if err != nil {
+			if errutil.IsNotFound(err) {
+				return c, string(clusterStatusNotFound), nil
+			}
+			return nil, "", err
+		}
+		if c.NewStatus == string(clusterStatusError) {
+			err = fmt.Errorf("vkcs_kubernetes_cluster is in an error state: %s", c.StatusReason)
+			return c, c.NewStatus, err
+		}
+		if c.NewStatus == string(clusterStatusReconciling) {
+			return c, string(nodeGroupStatusCreating), nil
+		}
+		return c, c.NewStatus, nil
+	}
+}
+
+func kubernetesNodeGroupDeleteStateRefreshFunc(client *gophercloud.ServiceClient, nodeGroupID string, clusterID string) retry.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		n, err := nodegroups.Get(client, nodeGroupID).Extract()
+		if err == nil {
+			return n, string(nodeGroupStatusDeleting), nil
+		}
+
+		if !errutil.IsNotFound(err) {
+			return nil, "", err
+		}
+
+		c, err := clusters.Get(client, clusterID).Extract()
+		if err != nil {
+			if errutil.IsNotFound(err) {
+				return c, string(clusterStatusNotFound), nil
+			}
+			return nil, "", err
+		}
+		if c.NewStatus == string(clusterStatusError) {
+			err = fmt.Errorf("vkcs_kubernetes_cluster is in an error state: %s", c.StatusReason)
+			return c, c.NewStatus, err
+		}
+		if c.NewStatus == string(clusterStatusReconciling) {
+			return c, string(nodeGroupStatusDeleting), nil
+		}
+		return c, c.NewStatus, nil
+	}
+}
```

Comment:
By the way, wouldn't it be the same if we wait for node group to be created/deleted at the first and then for the cluster status in different functions? This way we could add a simple function for waiting for node group `NOT_FOUND`/`RUNNING`.

Reply:
Rewrote to use separate wait function for node group